hasher-matcher-actioner/hmalib/common/mappings.py

```diff
@@ -1,37 +1,284 @@
+        return f"content_type:{full_class_name(content_type)}"
+
+
+class HMASignalTypeMapping(SignalTypeMapping):
+    """
+    An extension of the SignalTypeMapping defined in threatexchange.meta.
+
+    Enhancements:
+
+    1. State is pulled in from dynamodb when __init__ is first called.
+    2. Dedicated methods for getting signal and content_type by name. Instead of
+       depending on SignalTypeMapping's internal state.
+    """
+
+    @classmethod
+    def get_from_config_or_default(cls) -> "HMASignalTypeMapping":
+        """
+        Pull configs from HMAConfigs if available. If not, use defaults.
+        """
+        all_content_types = ToggleableContentTypeConfig.get_all()
+        all_signal_types = ToggleableSignalTypeConfig.get_all()
+
+        if len(all_content_types) + len(all_signal_types) == 0:
+            # We have never written content or signal types to database, so use
+            # default values instead.
+            (
+                all_content_types,
+                all_signal_types,
+            ) = cls._get_default_configs()
+
+        enabled_content_types = [
+            import_class(ct.content_type_class)
+            for ct in all_content_types
+            if ct.enabled
+        ]
+        enabled_signal_types = [
+            import_class(st.signal_type_class) for st in all_signal_types if st.enabled
+        ]
+
+        return HMASignalTypeMapping(enabled_content_types, enabled_signal_types)
+
+    @classmethod
+    def _get_default_configs(
+        cls,
+    ) -> t.Tuple[
+        t.List[ToggleableContentTypeConfig], t.List[ToggleableSignalTypeConfig]
+    ]:
+        """
+        Return default ToggleableContentTypeConfigs and
+        ToggleableSignalTypeConfigs. These are not guaranteed to be in the
+        config database.
+        """
+        content_type_configs = []
+        for _type in DEFAULT_SIGNAL_AND_CONTENT_TYPES.content_by_name.values():
+            content_config = ToggleableContentTypeConfig(
+                name=ToggleableContentTypeConfig.get_name_from_type(_type),
+                content_type_class=full_class_name(_type),
+            )
+            content_type_configs.append(content_config)
+
+        signal_type_configs = []
+        for _type in DEFAULT_SIGNAL_AND_CONTENT_TYPES.signal_type_by_name.values():
+            signal_config = ToggleableSignalTypeConfig(
+                name=ToggleableSignalTypeConfig.get_name_from_type(_type),
+                signal_type_class=full_class_name(_type),
+            )
+            signal_type_configs.append(signal_config)
+
+        return (content_type_configs, signal_type_configs)
+
+    def write_as_configs(self):
+        """
+        Write current state to dynamodb. This will create or update HMAConfigs
+        so that subsequent calls to get_from_config_or_default() return current
+        state.
+
+        Note: this will also disable signal and content types that are not part
+        of current state.
+        """
```

Comment:
Is the most common behavior that we write everything, or we'd write only single changes? I'm imagine a UI with a series of checkboxes

Reply:
I am not entirely sure how this would look in reality. I can't even remember why I wrote this method as it currently exists. You can't modify HMASignalTypeMapping any way. Let me rethink this bit.